src/aap_eda/services/project/scm.py

```diff
@@ -0,0 +1,283 @@
+            _executor = PlaybookExecutor()
+
+        vars = [f"project_path={path}"]
+        final_url = url
+        secret = ""
+        key_file = None
+        gpg_key_file = None
+        if credential:
+            inputs = inputs_from_store(credential.inputs.get_secret_value())
+            secret = inputs.get("password", "")
+            key_data = inputs.get("ssh_key_data", "")
+
+            final_url = cls.build_url(
+                url,
+                inputs.get("username", ""),
+                secret,
+                key_data,
+            )
+
+            if key_data:  # ssh
+                key_file = tempfile.NamedTemporaryFile("w+t")
+                key_file.write(key_data)
+                key_file.flush()
+                vars.append(f"key_file={key_file.name}")
+                key_password = inputs.get("ssh_key_unlock")
+                if key_password:
+                    cls.decrypt_key_file(key_file.name, key_password)
+
+        if gpg_credential:
+            gpg_inputs = inputs_from_store(
+                gpg_credential.inputs.get_secret_value()
+            )
+            gpg_key = gpg_inputs.get("gpg_public_key")
+            gpg_key_file = tempfile.NamedTemporaryFile("w+t")
+            gpg_key_file.write(gpg_key)
+            gpg_key_file.flush()
+            vars.append("verify_commit=true")
+            cls.add_gpg_key(gpg_key_file.name)
+
+        if not verify_ssl:
+            _executor.ENVIRON["GIT_SSL_NO_VERIFY"] = "true"
+
+        vars.append(f"scm_url={final_url}")
+        if branch:
+            vars.append(f"scm_branch={branch}")
+        if refspec:
+            vars.append(f"scm_refspec={refspec}")
+        if depth:
+            vars.append(f"depth={depth}")
+
+        vars_str = " ".join(vars)
+
+        cmd = ["-e", f"'{vars_str}'", "-t", "update_git"]
+        logger.info("Cloning repository: %s", url)
+        try:
+            git_hash = _executor(cmd)
+        except ScmError as e:
+            msg = str(e)
+            if secret:
+                msg = msg.replace(secret, "****", 1)
+            logger.warning("SCM clone failed: %s", msg)
+            raise ScmError(msg) from None
+        finally:
+            if key_file:
+                key_file.close()
+            if gpg_key_file:
+                gpg_key_file.close()
+        instance = cls(path, _executor=_executor)
+        instance.git_hash = git_hash
+        return instance
+
+    def _execute_cmd(self, cmd: Iterable[str]):
+        return self._executor(cmd, cwd=self.root)
+
+    @classmethod
+    def build_url(
+        cls, url: str, user: str, password: str, ssh_key: str
+    ) -> str:
+        result = urlparse(url)
+        domain = result.netloc.split("@")[-1]
+        scheme = result.scheme
+        path = result.path
+
+        if ssh_key and scheme == "" and path.startswith("git@"):
+            return url
+        elif user and password:
+            domain = f"{user}:{password}@{domain}"
```

Comment:
We need also to support only a token, without user. Ref: https://github.com/ansible/eda-server/blob/main/src/aap_eda/services/project/git.py#L143

There is a bug, that I can not find, where there are providers that have special characters in the user. It might be a good opportunity to set the url with url encoding. 

```
from urllib.parse import quote_plus
...
    encoded_user = quote_plus(user)
    encoded_password = quote_plus(password)
    domain = f"{encoded_user}:{encoded_password}@{domain}"
```
Ref: https://stackoverflow.com/questions/37841914/how-do-i-pass-username-and-password-while-using-ansible-git-module